### PR TITLE
Added abstraction for different backends/environments

### LINKED ISSUE #148: Use random file/blob name in the integration tests
Currently we use [constant file name](https://github.com/kubeflow/fairing/blob/master/tests/integration/test_kubeflow_training.py#L15) for test files that is causing random test failures.

### PR DESCRIPTION
**Pull request by karthikv2k**
Summary:
Currently we have a lot of configuration in a typical fairing job like preprocessor, builder, deployer etc. We added high level APIs recently to give Data Scientists a better way to do training and creating prediction endpoint without diving into much details. This PR makes it easy to switch between different environments like Kubeflow and other managed backends easily, just one line change. This will also help in consolidating a lot of default/guess logic into one place for each environment. Thus, we can provide much better experience based on what environment the user is in. For example, if user uses KF on GKE then we can mount GCP credentials by default. Another example is we can use K8s Job if running on bare K8s and use TF-Job if running on KF. Also fixes #148 .

TODOs in future PR:
#139 (Added integration tests for the TrainJob API but test for PredictionEndpoint is still pending and needs #144 ), #143 , Move all the guess/defaulting logic to appropriate backend classes

Example:
Deploying to a KF on GKE cluster:
```
backend = KubeflowGKEBackend()
train_job = TrainJob(train_fn, BASE_IMAGE, DOCKER_REGISTRY, backend)
train_job.submit()
```
Then only changing a line to deploy the same training job to GCP managed backend
```
backend = GCPManagedBackend()
train_job = TrainJob(train_fn, BASE_IMAGE, DOCKER_REGISTRY, backend)
train_job.submit()
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/fairing/144)

### CODE REVIEW

**Comment by karthikv2k** [comment]
/assign @r2d4 @rpasricha 
CC @kramachandran @abhi-g

**Comment by karthikv2k** [comment]
/hold
Making some changes.

File: fairing/backends/backends.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+class KubeflowGKEBackend(GKEBackend):
+
+    def __init__(self, namespace="kubeflow"):
+        super(KubeflowGKEBackend, self).__init__(namespace)
+    
+    def get_training_deployer(self):
+        return TfJob(namespace=self._namespace, pod_spec_mutators=[gcp.add_gcp_credentials_if_exists])
+
+class GCPManagedBackend(KubernetesBackend):
```
Comment (rpasricha) [inline fairing/backends/backends.py]: Should this class be extending from KubernetesBackend?

File: fairing/backends/backends.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        self._project_id = project_id or gcp.guess_project_name()
+        self._region = region or 'us-central1'
+        self._training_scale_tier =  training_scale_tier or 'BASIC'
+    
+    def get_training_deployer(self):
+        return GCPJob(self._project_id, self._region, self._training_scale_tier)
+
+    def get_serving_deployer(self, model_class):
+        # currently GCP serving deployer doesn't implement deployer interface
+        raise NotImplementedError("GCP managed serving is not implemented.")
```
Comment (rpasricha) [inline fairing/backends/backends.py]: We do have fairing.deployers.gcp.GCPServingDeployer

File: fairing/cloud/gcp.py [hunk 3]
```diff
@@ -2,10 +2,14 @@
...
 from fairing.constants import constants
 from kubernetes import client
+import logging
 
 import os
 import json
 
+logger = logging.getLogger(__name__)
+
+GCP_CREDS_SECRET_NAME = 'user-gcp-sa'
```
Comment (rpasricha) [inline fairing/cloud/gcp.py]: This should go into fairing/constants/constants.py

File: tests/integration/test_kubeflow_training.py [hunk 4]
```diff
@@ -36,87 +62,125 @@ def train_fn_with_gcs_access():
...
+            namespace='default',
+            gcs_prefix=temp_gcs_prefix,
+            capsys=capsys,
+            expected_result=None)
 
     msg = 'Unable to mount credentials: '\
           'Secret user-gcp-sa not found in namespace default'
     assert msg in str(err.value)
 
+def run_submission_with_high_level_api(backend, entry_point, capsys, expected_result):
```
Comment (rpasricha) [inline tests/integration/test_kubeflow_training.py]: The high level api tests should be in a separate file, especially as they'll get more complex over time.

File: fairing/backends/backends.py (hunk 2 above)
Comment (karthikv2k) [reply fairing/backends/backends.py]: Yes but we need it to support deployer interface for integration with high level api. Let's fix this in another PR where we can do some temporary hack till we get a container support. sounds good?

File: fairing/cloud/gcp.py (hunk 3 above)
Comment (karthikv2k) [reply fairing/cloud/gcp.py]: done.

File: tests/integration/test_kubeflow_training.py (hunk 4 above)
Comment (karthikv2k) [reply tests/integration/test_kubeflow_training.py]: done.

File: examples/prediction/xgboost-high-level-apis.ipynb [hunk 5]
```diff
@@ -187,6 +187,29 @@
...
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "from fairing import TrainJob\n",
+    "from fairing.backends import GCPManagedBackend\n",
+    "backend = GCPManagedBackend()\n",
+    "train_job = TrainJob(HousingServe, BASE_IMAGE, input_files=['ames_dataset/train.csv', \"requirements.txt\"],\n",
+    "                     docker_registry=DOCKER_REGISTRY, backend=backend)\n",
```
[the remark this answers is not in the document]
Comment (karthikv2k) [reply examples/prediction/xgboost-high-level-apis.ipynb]: Done.

**Comment by karthikv2k** [comment]
#118

**Comment by kramachandran** [comment]
Thanks Karthik. This looks great. 

So PM LGTM

**Comment by abhi-g** [comment]
Is this ready to merge?

File: fairing/backends/backends.py (hunk 2 above)
Comment (rpasricha) [reply fairing/backends/backends.py]: Sure, sounds great

**Comment by rpasricha** [comment]
/lgtm
/approve

Thanks for making the change!

**Pull request merged**